### ISSUE TITLE
[explicit-function-return-type] Return type needed after injection by higher order functions despite options

### ISSUE THREAD

**Issue by kace91** [typescript-eslint/typescript-eslint#758]
**Repro**

```JSON
"rules": {
    "@typescript-eslint/explicit-function-return-type": [
      "warn",
      {
        "allowTypedFunctionExpressions": true,
        "allowHigherOrderFunctions": true
      }
    ],
}
```

```TS


type MyFunctionType = (randomString: string) => string;
type InjectString<T> = (injectedString: string) => T;

const myfunction: InjectString<MyFunctionType> = injectedString => input => injectedString + input; // rule complains of missing return

const myFunctionAfterInjection = myfunction('potato'); // TS correctly infers a result of type MyFunctionType
const myFunctionResolved = myFunctionAfterInjection('anotherPotato'); // TS correctly infers a result of type string

```

**Expected Result**
Since 
a) my function is already typed
b)  I've marked as true both allowTypedFunctionExpressions and allowHigherOrderFunctions
c) Typescript is able to handle inference fine

I would expect the rule to accept this as an explicitly typed return type
**Actual Result**

The rule forces me to state the last return explicitly.

**Additional Info**

I could see how some people would prefer the last return function to be explicitly stated anyway, but I can't find a way to make my preference work: Typing just the function, but keeping the rule active for cases where the return type hasn't been stated in the function itself nor in the return type. I would like either dropping the need of the last return, a new option that handles this use case or a different rule if this is considered out of the scope for this specific rule - any way that can handle the use case would be fine.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.13.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `5.10.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
Note that this rule has nothing to do with helping typescript infer a type, or even a correct type. The intention behind the rule is that you explicitly declare a return type contract within your function declarations, so that your code is clear.

The higher order function support is intended to save you double typing functions when chaining them together - you only have to type the last return type because the rest of the return types are clear, and explicitly typing them is confusing to read, and entirely redundant.:
```ts
// type only the last return
const x = () => (arg1: string) => (arg2: number): string => 'foo' 
// vs
// type every return
const y = () : ((arg1: string) => (arg2: number) => string) =>
  (arg1: string): ((arg2: number) => string) =>
  (arg2: number): string => 'foo'
```

Potentially could add support for `allowTypedFunctionExpressions` within the `allowHigherOrderFunctions`.
```
// does not error
const x: HigherOrderType = () => (arg1: string) => (arg2: number) => 'foo' 
```

**Comment by kace91** [comment]
> The higher order function support is intended to save you double typing functions when chaining them together - you only have to type the last return type because the rest of the return types are clear, and explicitly typing them is confusing to read, and entirely redundant.:

I'm aware of that, sorry if I wasn't clear (English isn't my first language).

 I didn't expect the higher order function rule to let me skip the return value by itself - my issue was more that I expected that rule to "combine" with the other (`allowTypedFunctionExpressions`) when both are active so that it ignores the requirement of stating the return value explicitly for the whole chain.

 The reasoning for this in my mind is that, if the rule `allowTypedFunctionExpressions` saves you from double typing what's already stated in the function definition, and `allowHigherOrderFunctions` helps readability by treating the whole curried chain as a single function (requiring only the last return value), then it would follow that the combination would also skip the need for the return value of the chain when the type has already been defined in the function declaration.

That is to say, if a user wants to type :

```TS
const notInjectedFunction: MyFunctionType = randomstring => randomstring.toUpperCase();
```
without specifying a return value (by setting the rule `allowTypedFunctionExpressions` to true), then I think it would be consistent to assume that they would also want to skip stating the type in 
```TS
const myfunction: InjectString<MyFunctionType> = injectedString => input => injectedString + input;
```
Although having that be a option for the user is also great.

**Comment by bradzacher** [comment]
Sorry, wrong tag - i meant that the two options should work together.

**Comment by bradzacher** [comment]
happy to accept a PR if you think this change is valuable to you.
